File: typescript/src/adapters.ts
export interface ExecutableTool<TResult = string> {
  name: string;
  description?: string;
  inputSchema: Record<string, unknown>;
  execute(input?: Record<string, unknown>): Promise<TResult>;
}

export interface AISDKToolFactory<TTool = unknown> {
  (options: {
    description?: string;
    inputSchema: Record<string, unknown>;
    execute: (input: Record<string, unknown>) => Promise<string>;
  }): TTool;
}

export function toAISDKTools<TTool = unknown>(
  tools: Record<string, ExecutableTool<string>>,
  options: { tool?: AISDKToolFactory<TTool> } = {},
): Record<string, TTool | Omit<ExecutableTool<string>, "name">> {
  const result: Record<string, TTool | Omit<ExecutableTool<string>, "name">> = {};
  for (const [name, executable] of Object.entries(tools)) {
    const definition = {
      description: executable.description,
      inputSchema: executable.inputSchema,
      execute: (input: Record<string, unknown>) => executable.execute(input),
    };
    result[name] = options.tool ? options.tool(definition) : definition;
  }
  return result;
}

export function toMastraTools(
  tools: Record<string, ExecutableTool<string>>,
): Record<string, Omit<ExecutableTool<string>, "name">> {
  const result: Record<string, Omit<ExecutableTool<string>, "name">> = {};
  for (const [name, executable] of Object.entries(tools)) {
    result[name] = {
      description: executable.description,
      inputSchema: executable.inputSchema,
      execute: (input: Record<string, unknown> = {}) => executable.execute(input),
    };
  }
  return result;
}

File: python/mcp-compressor/mcp_compressor/adapters.py
from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import Any, TypeVar

from mcp_compressor.client import ExecutableTool

TTool = TypeVar("TTool")
ToolFactory = Callable[[dict[str, Any]], TTool]


def to_ai_sdk_tools(
    tools: Mapping[str, ExecutableTool],
    *,
    tool: ToolFactory | None = None,
) -> dict[str, Any]:
    """Convert executable tools into AI-SDK-style tool definitions.

    The optional ``tool`` factory mirrors the TypeScript adapter: pass a framework helper
    if you want framework-specific wrappers, or omit it to get plain structural objects.
    """
    result: dict[str, Any] = {}
    for name, executable in tools.items():
        definition = {
            "description": executable.description,
            "input_schema": executable.input_schema,
            "execute": executable.execute,
        }
        result[name] = tool(definition) if tool else definition
    return result


def to_mastra_tools(tools: Mapping[str, ExecutableTool]) -> dict[str, dict[str, Any]]:
    """Convert executable tools into a Mastra-like structural tool map."""
    return {
        name: {
            "description": executable.description,
            "input_schema": executable.input_schema,
            "execute": executable.execute,
        }
        for name, executable in tools.items()
    }
